perf/src/client.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    net::{IpAddr, Ipv6Addr, SocketAddr},
+    sync::Arc,
+};
+
+use anyhow::{Context, Result};
+use bytes::Bytes;
+use futures::StreamExt;
+use structopt::StructOpt;
+use tokio::sync::Semaphore;
+use tracing::{error, info, trace};
+
+/// Connects to a QUIC perf server and maintains a specified pattern of requests until interrupted
+#[derive(StructOpt)]
+#[structopt(name = "client")]
+struct Opt {
+    /// Address to connect to
+    #[structopt(default_value = "localhost:4433")]
+    host: String,
+    /// Number of unidirectional requests to maintain concurrently
+    #[structopt(long, default_value = "0")]
+    uni_requests: u64,
+    /// Number of bidirectional requests to maintain concurrently
+    #[structopt(long, default_value = "1")]
+    bi_requests: u64,
+    /// Number of bytes to request
+    #[structopt(long, default_value = "1048576")]
+    request_size: u64,
+}
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() {
+    let opt = Opt::from_args();
+
+    tracing_subscriber::fmt::init();
+
+    if let Err(e) = run(opt).await {
+        error!("{:#}", e);
+    }
+}
+
+async fn run(opt: Opt) -> Result<()> {
+    let addr = tokio::net::lookup_host(&opt.host)
+        .await
+        .context("resolving host")?
+        .next()
+        .unwrap();
+    let host_name = opt.host.split(':').next().unwrap();
+
+    info!("connecting to {} at {}", host_name, addr);
+
+    let endpoint = quinn::EndpointBuilder::default();
+
+    let (endpoint, _) = endpoint
+        .bind(&SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0))
+        .context("binding endpoint")?;
+
+    let mut cfg = quinn::ClientConfigBuilder::default();
+    cfg.protocols(&[b"perf"]);
+    let mut cfg = cfg.build();
+    let tls_cfg: &mut rustls::ClientConfig = Arc::get_mut(&mut cfg.crypto).unwrap();
+    tls_cfg
```

Comment:
Maybe move this behind an `--insecure` flag? There can definitely be use cases to run this against places with a valid cert. For those use-cases it might also be nice to either
- specify the ip independent from the hostname (similar to how `curl --resolve` works)
- allow to overwrite the SNI

I would recommend the first option. But this is also fine for a follow-up, and i'm happy to add it.